### PR TITLE
Allow password encodings in config

### PR DESCRIPTION
**Pull request by tobes**
From #415 this adds universal base64 encoding via
```
module {
   value:base64 = '<encoded>'
}
```

@ultrabug if you are happy with this concept I'll update this PR with some documentation

### CODE REVIEW

File: py3status/parse_config.py [hunk 1]
```diff
@@ -441,6 +442,12 @@ def parse(self, dictionary=None, end_token=None):
                     dictionary[name] = value
                 # assignment of value
                 elif t_value == '=':
+                    # if we have a colon in the name of a setting then it
+                    # indicates that it has been encoded.
+                    if ':' in name:
+                        (name, scheme) = name.split(':')
+                        if scheme == 'base64':
```
Comment (ultrabug) [inline py3status/parse_config.py]: if this ought to be extendable, I'd rather have this done in a dedicated method please

File: py3status/parse_config.py [hunk 2]
```diff
@@ -100,7 +101,7 @@ class ConfigParser:
         '|(?P<literal>'
         r'("(?:[^"\\]|\\.)*")'  # double quoted string
         r"|('(?:[^'\\]|\\.)*')"  # single quoted string
-        '|([a-z_][a-z0-9_]*)'  # token
+        '|([a-z_][a-z0-9_]*(:[a-z0-9_]+)?)'  # token
```
Comment (ultrabug) [inline py3status/parse_config.py]: I think `a-z0-9_` equivalent is `\w`

File: py3status/parse_config.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 import glob
 import os
 import imp
+import base64
```
Comment (ultrabug) [inline py3status/parse_config.py]: bad kitty

**Comment by ultrabug** [comment]
I pretty like the concept of generalization here yes tho I'm still not convinced that base64 encoding anything increases the security of #415

**Comment by tobes** [comment]
> I'm still not convinced that base64 encoding anything increases the security

I was off the same opinion but it is still useful.  It means I can look at you config and not just read your password by accident.  But it is completely insecure

We could do something proper secure and have py3status prompt for a pass phrase and not start affected modules till it has one.  But I'm not sure that it is worth the effort at the moment.

File: py3status/parse_config.py (hunk 2 above)
Comment (tobes) [inline py3status/parse_config.py]: sometimes explicit is good eg is `-` in `\w`

Comment (ultrabug) [inline py3status/parse_config.py]: true

**Comment by tobes** [comment]
I've moved the processing into it's own method.  It means we can now do some insane stuff in our configs, not that it should be supported.

```
static_string test {
    format:base64 = 'MTIzNA=='
    format += '=='
    format:base64 += 'MTIzNA=='
}
```

File: doc/README.md [hunk 4]
```diff
@@ -7,6 +7,7 @@ py3status documentation
 * [Available modules](#available_modules)
 * [Ordering modules](#ordering_modules)
 * [Configuring modules](#configuring_modules)
+* [Obfuscation](#obfuscation)
```
Comment (ultrabug) [inline doc/README.md]: **Configuration obfuscation** maybe more understandable ?

File: doc/README.md [hunk 5]
```diff
@@ -83,14 +84,34 @@ imap {
...
     user = 'mylogin'
     on_click 1 = "exec thunderbird"
 }
 ```
 
+#### <a name="obfuscation"></a>Obfuscation
+
+__New in version 3.1__
+
+In the previous example configure the users password is in plain text.  Users may
```
Comment (ultrabug) [inline doc/README.md]: ... example configuration ...

**Comment by tobes** [comment]
Made those doc fixes

**Comment by tobes** [comment]
So I was having a think about the whole privacy stuff and decided that it was an interesting topic.

py3status would easily leak any information that you encode eg if you use logging, notifications, put it in your format string.  So I've updated this PR to stop this leakage.  It is not an easy problem as we need to ensure that modules work, especially if they use external libs eg requests.

This implementation allows the value of 'private' config options only to be known to modules with the same name as the config option is given for.  And when used outside this `***` is used as the value.

All unicode methods should be available and work both when valid and not.

The whole thing is academic as all I need is to add a module that monkey patches this out and I can get the real values.  So this is more about keeping things semi-private and a bit of fun.

As well as `:base64` I have added `:hide` that is plain text in the config but will not leak to logs, i3bar etc

**Comment by tobes** [comment]
A few more updates.

* We are more resistant to modules trying to change the Private classes

* Only allow this for py3status modules

* Only allow for strings

* Check the 'scheme' exists

**Comment by ultrabug** [comment]
I'm tempted to use `:private` instead of `:hide`

Also, since :base64 et al are used to obfuscate, why don't we hide them too by default ?

**Comment by ultrabug** [comment]
> I'm tempted to use :private instead of :hide

dunno, :hide seems more explicit after some more thoughts

**Comment by tobes** [comment]
> Also, since :base64 et al are used to obfuscate, why don't we hide them too by default ?

We do the :hide is just so I can have my password hidden in the logs all :xxx are hidden in this way.

Do I need to make the docs clearer?

**Comment by tobes** [comment]
Hopefully clearer documentation

**Comment by ultrabug** [comment]
Thanks this is a neat addition :)

**Pull request merged**